### PR TITLE
Update CCD to spec

### PR DESCRIPTION
**Pull request by ericmarkmartin**
Add `crossOrigin` field to `CollectedClientData` and capture unknown keys in an extra field.

Additionally updates some formatting left over from credBlob.

Implements #52.

- [x] cargo fmt has been run
- [x] cargo test has been run and passes
- [x] documentation has been updated with relevant examples (if relevant)

### CODE REVIEW

File: src/proto.rs [hunk 1]
```diff
@@ -970,9 +961,17 @@ pub struct CollectedClientData {
     pub challenge: Base64UrlSafeData,
     /// The rp origin as the browser understood it.
     pub origin: String,
+    /// The inverse of the sameOriginWithAncestors argument value that was
+    /// passed into the internal method.
+    #[serde(rename = "crossOrigin", skip_serializing_if = "Option::is_none")]
+    pub cross_origin: Option<bool>,
```
Comment (Firstyear) [inline src/proto.rs:967]: Do we need to perform any validation around the option and it's usage? We should understand what it does ....

**Review by Firstyear — APPROVED** [review]
Looks good, just want to understand what cross_origin does, and if we need to validate anything about it.

File: src/proto.rs (hunk 1 above)
Comment (ericmarkmartin) [reply src/proto.rs:967]: I think it indicates if the javascript context that created the public key credential had a different origin than the origin that the credential is created for.

We could consider adding a predicate `allow_cross_origin` in the `WebauthnConfig` trait (defaulting to `false` always) that the user can set to indicate whether they want to accept such credentials.

**Comment by ericmarkmartin** [comment]
So I think `cross_origin` just indicates that the credential was created by javascript running on an origin different than that of the RP the credential was made for. I think it might be worth allowing users to disallow this behavior, but interested in your thoughts.

**Comment by Firstyear** [comment]
I think that sounds correct, we should have an allow_cross_origin bool in the config trait. By default we disallow it?

**Comment by ericmarkmartin** [comment]
Yeah I'd think so--I'll push new changes my tomorrow

File: src/core.rs [hunk 2]
```diff
@@ -297,7 +297,10 @@ impl<T> Webauthn<T> {
         }
 
         // Verify that the value of C.origin matches the Relying Party's origin.
-        if data.client_data_json.origin != self.config.get_origin() {
+        if !self.config.allow_cross_origin()
```
Comment (Firstyear) [inline src/core.rs:300]: I think this is not correct. https://www.w3.org/TR/webauthn-2/#sctn-verifying-assertion does not actually inspect crossOrigin during the assertion or registration, and we still need to check the origin is our rp origin.

So this makes me question, what is even the point of this boolean if we never inspect it during a registration/assertion? 

@agrinman Thoughts?

Comment (ericmarkmartin) [reply src/core.rs:300]: I think there's a bit [here](https://www.w3.org/TR/webauthn-2/#clientdatajson-verification) about limited verification if you cannot support a full JSON parser?

Comment (ericmarkmartin) [reply src/core.rs:300]: > Verifiers may use the following algorithm to verify an encoded CollectedClientData if they cannot support a full JSON parser

Comment (Firstyear) [reply src/core.rs:300]: Step 4 there still states:

```
A string, origin, that contains the expected origin that issued the request to the user agent.
```

The issue is that your logic as it's set, if you allow cross_origin, then the origin value isn't checked. This is incorrect.

The code should be:

```
if data.client_data_json.origin != self.config.get_origin() {
    // as before
}

if !self.config.allow_cross_origin() && data.client_data_json.cross_origin.unwrap_or(false) {
    // return error, dis allowing cross origin ccd
}
```

Comment (ericmarkmartin) [reply src/core.rs:300]: That makes sense

**Pull request merged**